### ISSUE TITLE
Learning Rate adjustment does not work in Pytorch>2.0

### ISSUE THREAD

**Issue by dh5414** [MIC-DKFZ/nnUNet#1781]
when you use Learning Rate adjustmen  in Pytorch>2.0,it can not work,because pytorch>2.0 has  detected call of `lr_scheduler.step()` before `optimizer.step()`.
To solve Problem,you should rewrite **nnUNetTrainerCosAnneal** Class and rewrite function:**on_train_epoch_start(self)** and **train_step(self, batch: dict) -> dict:**
adjust call order for`lr_scheduler.step()` and `optimizer.step()`
this code could be a refrence:
```python
from torch.optim.lr_scheduler import CosineAnnealingLR
from nnunetv2.training.nnUNetTrainer.nnUNetTrainer import *

class nnUNetTrainerCosAnneal(nnUNetTrainer):
    def configure_optimizers(self):
        optimizer = torch.optim.SGD(self.network.parameters(), self.initial_lr, weight_decay=self.weight_decay,
                                    momentum=0.99, nesterov=True)
        lr_scheduler = CosineAnnealingLR(optimizer, T_max=self.num_epochs)
        return optimizer, lr_scheduler
    def on_train_epoch_start(self):
        self.network.train()
        # self.lr_scheduler.step()
        self.print_to_log_file('')
        self.print_to_log_file(f'Epoch {self.current_epoch}')
        self.print_to_log_file(
            f"Current learning rate: {np.round(self.optimizer.param_groups[0]['lr'], decimals=5)}")
        # lrs are the same for all workers so we don't need to gather them in case of DDP training
        self.logger.log('lrs', self.optimizer.param_groups[0]['lr'], self.current_epoch)

    def train_step(self, batch: dict) -> dict:
        data = batch['data']
        target = batch['target']

        data = data.to(self.device, non_blocking=True)
        if isinstance(target, list):
            target = [i.to(self.device, non_blocking=True) for i in target]
        else:
            target = target.to(self.device, non_blocking=True)

        self.optimizer.zero_grad(set_to_none=True)
        # Autocast is a little bitch.
        # If the device_type is 'cpu' then it's slow as heck and needs to be disabled.
        # If the device_type is 'mps' then it will complain that mps is not implemented, even if enabled=False is set. Whyyyyyyy. (this is why we don't make use of enabled=False)
        # So autocast will only be active if we have a cuda device.
        with autocast(self.device.type, enabled=True) if self.device.type == 'cuda' else dummy_context():
            output = self.network(data)
            # del data
            l = self.loss(output, target)

        if self.grad_scaler is not None:
            self.grad_scaler.scale(l).backward()
            self.grad_scaler.unscale_(self.optimizer)
            torch.nn.utils.clip_grad_norm_(self.network.parameters(), 12)
            self.grad_scaler.step(self.optimizer)
            self.grad_scaler.update()
        else:
            l.backward()
            torch.nn.utils.clip_grad_norm_(self.network.parameters(), 12)
            self.optimizer.step()
        self.lr_scheduler.step()
        return {'loss': l.detach().cpu().numpy()}


```

**Comment by dojoh** [comment]
Hey,
thanks for the feedback! I will bring this to @FabianIsensee s attention!
Cheers.

**Comment by FabianIsensee** [comment]
Hey can you please provide information for why it doesn't work the way we do it? We step the lr_scheduler once per epoch (which is btw an error in your script because you step it every iteration but T_max is self.num_epochs!) in on_train_epoch_start
Do you have a link where we can read about this?

**Comment by dh5414** [comment]
> Hey can you please provide information for why it doesn't work the way we do it? We step the lr_scheduler once per epoch (which is btw an error in your script because you step it every iteration but T_max is self.num_epochs!) in on_train_epoch_start Do you have a link where we can read about this?

Thank you for pointing out the mistake in T_max setting."
